libraries/Microsoft.Bot.Builder.Azure/CosmosDbSqlStorage.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.Documents.Linq;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// Models IStorage around a CosmosDB SQL (DocumentDB)
+    /// </summary>
+    public class CosmosDbSqlStorage : IStorage
+    {
+        private readonly string databaseId;
+        private readonly string collectionId;
+        private readonly DocumentClient client;
+
+        private const string UserAgentSuffix = "Microsoft-BotFramework 4.0.0";
+
+        private static JsonSerializer jsonSerializer = JsonSerializer.Create(new JsonSerializerSettings()
```

Comment:
The same could be accomplised if the State class properties are decoreated with `JsonProperty`.

Using the `CamelCasePropertyNamesContractResolver` is not enough since it only serializes to pascalCase, but de-serializes JSON into JObject as-is... So using `dynamic` objects causes issues. The C# property may be `Title`, gets serialized to `title` and the de-serialized JObject will still be `title`.
For the Serializer to handle this automatically, a custom converter is required and must be thoroughly tested ([Relevant discussion](https://stackoverflow.com/questions/9247478/pascal-case-dynamic-properties-with-json-net)).

I'll take a better look at this later on and publish a new PR when it is thoroughly tested. Sound ok?